### ISSUE TITLE
Change Chromium cask to track last-known good releases instead of trunk

### ISSUE THREAD

**Issue by wfdd** [caskroom/homebrew-cask#15357]
The Chromium cask is tracking Chromium trunk, which is bleeding-profusely-edge; these builds are often very buggy and are—basically—unusable. It would make more sense to rename the current Chromium cask to `chromium-trunk` (and ship it off to [caskroom/homebrew-versions](https://github.com/caskroom/homebrew-versions)) or somesuch and create a new cask for Chromium that would mirror Chrome releases.

**Comment by adityadalal924** [comment]
Can you give some more information about this issue?

Where can "versioned" Chromium releases be found (that mirror Chrome releases)?

I believe you are looking for something like http://www.freesmug.org/chromium, which you can install with [`brew cask install freesmug-chromium`](https://github.com/caskroom/homebrew-cask/blob/master/Casks/freesmug-chromium.rb)

**Comment by wfdd** [comment]
Right, but FreeSMUG is not an official build; apparently, they bundle Sparkle—though the release notes has 'this FreeSMUG build hasn't [got the] "Check for Updates…" command' since at least January—and Flash, which I don't want.

All that's required here is to link to the build the latest Chrome stable release derives from. Naturally, we're gonna have to (a) figure out how to locate that build and (b) update the cask with each new Chrome release.

Basically, I was throwing it out as an idea before I get down to it.

**Comment by adityadalal924** [comment]
Homebrew Cask does not build from source, we are only a distribution mechanism for already available packages (ie, .dmg or .pkg that installs an .app).

Feel free to reopen this issue if you can find an official source of Chromium builds that fits your requirements, but until then, closing the issue as it's out of scope for the project.

Thanks for the suggestion, anyway.

**Comment by webdrone** [comment]
You do not need to build from source, just link to the latest built Chromium package which correspond to the latest stable Chrome release. This can be done by looking up the Chrome version and finding the corresponding Chromium one from the list of built available packages to link to.

**Comment by adityadalal924** [comment]
@webdrone Thanks for the additional information. So, in this case, the current stable version would be `46.0.2490.86`, with a `branch_base_position` of `344925`.

I then see a `chrome-mac.zip`, but there seems to be a `generation` URL parameter, without which the download will not work.

The way I could see this working is if we have the version stanza be `344925-1440204568655000`, ie. base_branch_position-generation, the two things required for the url.

Pinging @vitorgalvao for his thoughts.

**Comment by wfdd** [comment]
It does actually appear to work without the generation thingy; it doesn't work if you take `media` out of the query string.

**Comment by vitorgalvao** [comment]
No.

Why would we be dependent on an external unofficial<sup>1</sup> source (`omahaproxy.appspot.com`) to update a cask officially?

As @paulirish mentioned (and he’d know), “[there is no such thing as a stable build of Chromium](https://github.com/caskroom/homebrew-cask/issues/6747#issue-46135169)”, **that’s the whole point of Chromium**. It makes no sense for us to pick an arbitrary<sup>2</sup> release number *based on a source we have no reason to trust blindly* and freeze it there.

---
<sup>1</sup> I’m guessing, but I see nothing there that makes it look official in any way.

<sup>2</sup> And yes, it is arbitrary to the people who don’t know we’re doing it. We follow user expectations, and a Chromium user should expect the latest build.

**Comment by adityadalal924** [comment]
@vitorgalvao The URL does seem official, according to https://www.chromium.org/administrators/frequently-asked-questions, specifically

```
How do I know what the most current version of Google Chrome is for Windows?
    We've created a utility at https://omahaproxy.appspot.com/win to list the current stable version on Windows.

How do I know what the most current version of Google Chrome is for other platforms?
    As above, https://omahaproxy.appspot.com/viewer provides a list of current revisions.
```

but I do agree that pulling from trunk (the current approach) seems to be the "desired" outcome.

Re having a stable build, you are welcome to create and use "chromium tracking chrome" Cask, which could potentially live in `homebrew-unofficial` or `homebrew-versions`, or of course, your own personal tap.

**Comment by vitorgalvao** [comment]
Oh, right, in these cases I always mean to mention the possibility of personal taps, but keep forgetting. Thank you for pointing it out.

**Comment by wfdd** [comment]
Your 'no' response appears to rest on the following premises:

1. That the source is unofficial, which it isn't, as has already been pointed out.
2. That the version source is the same as the build source, which it isn't. (To be clear, both the build and version sources are official.)
3. That the build is not actually stable by virtue of the fact that there's no Chromium 'stable' channel; or that all Chromium builds are equally stable or unstable.

None of these are true.

**Comment by vitorgalvao** [comment]
> That the source is unofficial, which it isn't, as has already been pointed out.

I pointed out in the original post that I was guessing from what it looked like. That was the lesser point, though, and since I’ve already been corrected but didn’t touch that point again, you can see that wasn’t the most important reason (as I hoped was clear by the amount of detail that went into the other part of the question).

> That the version source is the same as the build source, which it isn't. (To be clear, both the build and version sources are official.)

No, I never made that point. My use of the word “source” was referring to `omahaproxy.appspot.com`, as in “the source from where we’d be getting the relevant information”.

> That the build is not actually stable by virtue of the fact that there's no Chromium 'stable' channel; or that all Chromium builds are about equally stable or unstable.

I also never made that point or even hinted at that.

> None of these are true.

And I never made those points (except for the first point, which is the least relevant, and I’ve already addressed). In fairness though, you did say my response *appears* to rest on those premisses. It doesn’t.

Also, forgot to address this before:

> (I think there's been some talk about Cask cron jobs lately, though I don't recall where.)

I believe I know what you’re referring to, in the [caskroom/versions](https://github.com/caskroom/homebrew-versions) repo, but I’m not currently finding it. But no, there hasn’t been any serious talk of anything that would apply specifically to this case. What was discussed were automated updates specifically for nightly builds, or updates using official appcasts.

**Comment by wfdd** [comment]
Quoting from earlier:

> And yes, it is arbitrary to the people who don’t know we’re doing it. We follow user expectations, and a Chromium user should expect the latest build.

Who is this imaginary Chromium user who'd wanna run a build that crashes whenever you press the `cmd` key in the address bar, for instance?  And if the only stumbling block is that the name should be reserved for Chromium CI, we don't have to follow through with repurposing this *specific* cask; we could create a new one.

Would you like to offer a usable, open-source alternative to Chrome? A Chromium cask tracking Chrome stable releases, I think, would be it. It is what several Linux distros do (though they do also patch it up, mostly to address Linux-specific issues with Chromium).

**Comment by paulirish** [comment]
A few things:

* Omahaproxy is run by Chrome Infra team. It's totally canonical.
* download-chromium.appspot.com is run by Chromium developers (Francois Beaufort and me), and is a reliable source for getting a download link. [its repo](https://github.com/beaufortfrancois/download-chromium), btw.
* The earlier calculation for getting a stable build is incorrect, unfortunately. The `branch_base_position` points to where that milestone (Chrome 47 or 48) was branched from the Chromium master (development) branch. As the branch is stabilized, it's `branch_base_position` does not change, but its `branch_commit` does.
* The builds available from the `commondatastorage.googleapis.com` location only include builds from master branch and do not include stabilized branch builds.  

As such, there is no official repo of chromium binaries tracking the stable release branches. Sorry!
It's possible to automate building them when stable branch updates, but I haven't seen anyone do it yet.

http://chromium-status.appspot.com/revisions lists recent "lkgr" last-known-good-revision. This will be the revision where all tests passed AND no regressions were detected. This is a better build for the homebrew audience, IMO.

**Comment by wfdd** [comment]
Thanks for the insight. That does sound like a much better option.

**Comment by vitorgalvao** [comment]
@wfdd Why do people that seldom (if ever) contribute always get so insulted and insistent when we say “no” to a request? [Homebrew taps](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/brew-tap.md) are a wonderful system. It works great, and allows immense flexibility. If you think your cask request has so much merit and deserves to exists *after it has been explained why we won’t support it officially*, then make a tap! It’s incredibly easy to do so. It’s not like non-inclusion in the main project stops you from actually having it. But no, people that do what you do never want to be the ones to actually do the maintenance, homebrew-cask team’s time and resources be damned, as we must work for you. Demands ever present, but offers to do the actual legwork, that you won’t do.

> Would you like to offer a usable, open-source alternative to Chrome?

No, I’d like to offer *exactly* what the developers offer. Whatever that is, I trust they have their reasons. Want something else? Ask the official team! So many times we get requests for workarounds and specific things that should be taken care of upstream. No wonder [all of programming sucks](http://www.stilldrinking.org/programming-sucks) and is hanging on thin threads. Instead of fixing things at the source, everyone wants to build workarounds on the part of the system they use. Again, asking upstream is the way, especially since asking the official team means you’ll get an official answer as to why the builds aren’t distributed or distributed in the way you want or expect, and that insight is extremely valuable.

> As such, there is no official repo of chromium binaries tracking the stable release branches.

And as such, we won’t provide them, as well. The mentioned last-known-good-revision site looks good in theory, but there are multiple releases a day, which is untenable with the system homebrew and homebrew-cask use, there’s no way we’d be able to keep up (not so with the official chromium cask, that always tracks the latest link). The only way this could work is if that `lkgr` repo itself provided the cask. That way the team/person behind it, could automate the adding of the build with the update of the cask.

Thank you for the fast and detailed response, @paulirish. Getting information directly from upstream is always welcome and appreciated.

**Comment by paulirish** [comment]
> The only way this could work is if that lkgr repo itself provided the cask. 

We can do this for you. I opened https://github.com/beaufortfrancois/download-chromium/issues/8 to discuss it.. 

Once done, it will probably end up being a 5-character diff to your cask formula. :)

**Comment by adityadalal924** [comment]
On hold until https://github.com/beaufortfrancois/download-chromium/issues/8 is resolved

**Comment by vitorgalvao** [comment]
Not sure it was clear from my post, but what I mean by `lkgr` providing the cask is they would also host it. This is a must, because when we have [this](http://i.imgur.com/8vgeFtO.png) (multiple daily builds), submitting it automatically to this repo would be untenable. Though we’re pretty quick with merging pull requests, those would quickly overwhelm us. We support direct linking to a cask, so a user could do something like `brew cask install https://chromium-status.appspot.com/revisions/chromium.rb`, and it’d work.

This would be great once we support `upgrade` functionality.

If the official chromium team so desired, I’d even have no quarrel with removing our current cask and recommending yours as the official way to get chromium through homebrew-cask. Maintaining it as is is also not an issue, naturally.

Closing again since there is nothing left for us to do with this issue. Either the official team provides a link to the `lkgr` link or they don’t, but our cask will remain as is (or be deleted if the chromium team so desires).

**Comment by wfdd** [comment]
> Why do people that seldom (if ever) contribute always get so insulted and insistent when we say “no” to a request?

Where did it seem that I got insulted? You're not the only who's doing something for free and gets no thanks, unfortunately; we've all got something going on. But my gratitude for the work you do here is implied, and it is now also stated. I put the idea forward, I insisted on it a little, you're pretty adamant it's not happening (at least for now), and that is that.

**Comment by adityadalal924** [comment]
@vitorgalvao Since we support `:latest` and `:no_check`, would the below be an acceptable compromise (as many other Casks also use redirects)

```
cask :v1 => 'chromium' do
  version :latest
  sha256 :no_check

  # appspot.com is the official download host per the vendor homepage
  url 'https://download-chromium.appspot.com/dl/Mac/lkgr'
  name 'Chromium'
  homepage 'https://www.chromium.org/Home'
  license :oss

  app 'chrome-mac/Chromium.app'
end
```

where `https://download-chromium.appspot.com/dl/Mac/lkgr` should point to the required build.

**Comment by vitorgalvao** [comment]
@adityadalal924 Yes, that is a very good solution. Hadn’t crossed my mind.

**Comment by vitorgalvao** [comment]
However, I’ll also mention that `lkgr` being hosted officially is still a superior option, because of `upgrade`. When we support it, `upgrade` won’t touch `:latest` casks (because it has no way to know they were updated), while a versioned cask automatically updated would ensure people are just a `brew cask upgrade` away from having the latest good build of chromium.

**Comment by beaufortfrancois** [comment]
I've updated https://download-chromium.appspot.com website to include LKGR builds.
You can now for instance get the LKGR build of Chromium for Android at https://download-chromium.appspot.com/?platform=android&type=continuous or directly download it via https://download-chromium.appspot.com/dl/android?type=continuous
